Bug: Slower movement in higher mouse hertz

Hi,

Just found a bug related to hz in the mouse.

I was using the cable which only get to about 1000hz, the graph and sensitivity were quickly changing.

Now, after changing to 4000hz, all seem slowed down in the sensitivity of the graph. Might require to detect change in the hz or have a way to input your current hz, so you can do the calculations based on it.

It's like with the same or more mouse movement, the dots in the graph move less. From simply going from 1000 to 4000hz.

Changing the offset to half, got close to what it was. But it feels weird, as it takes longer reach the accel. Since the dots takes longer to move. Feels inconsistent.

---

This need to be fixed, otherwise, it's unusable for me at least.

Why does it guard against it?

```
❯ make debug_install
/usr/bin/make CC=gcc EXTRA_CFLAGS=-DDEBUG -C `pwd`/driver
make[1]: Entering directory '/home/user/Downloads/maccel/driver'
/usr/bin/make -C /lib/modules/6.13.2-2-cachyos/build M=/home/user/Downloads/maccel/driver
make[2]: Entering directory '/usr/lib/modules/6.13.2-2-cachyos/build'
make[3]: Entering directory '/home/user/Downloads/maccel/driver'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: clang version 19.1.7
  You are using:           gcc (GCC) 14.2.1 20250207
  CC [M]  maccel.o
gcc: error: unrecognized command-line option ‘-mretpoline-external-thunk’
gcc: error: unrecognized command-line option ‘-fsplit-lto-unit’
gcc: error: unrecognized command-line option ‘-fdebug-info-for-profiling’
gcc: error: unrecognized command-line option ‘-mllvm’
gcc: error: unrecognized command-line option ‘-mllvm’
gcc: error: unrecognized command-line option ‘-improved-fs-discriminator=true’
gcc: error: unrecognized command-line option ‘-fbasic-block-sections=labels’
make[5]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/scripts/Makefile.build:196: maccel.o] Error 1
make[4]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/Makefile:1997: .] Error 2
make[3]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/Makefile:251: __sub-make] Error 2
make[3]: Leaving directory '/home/user/Downloads/maccel/driver'
make[2]: *** [Makefile:251: __sub-make] Error 2
make[2]: Leaving directory '/usr/lib/modules/6.13.2-2-cachyos/build'
make[1]: *** [Makefile:7: build] Error 2
make[1]: Leaving directory '/home/user/Downloads/maccel/driver'
make: *** [Makefile:16: build] Error 2
```

It's a Lamzu 4k mini.

> Why does it guard against it?
> 
Idk, carried over from when I ported leetmouse. I probably don't need it.
> ```
> ❯ make debug_install
> /usr/bin/make CC=gcc EXTRA_CFLAGS=-DDEBUG -C `pwd`/driver
> make[1]: Entering directory '/maccel/driver'
> /usr/bin/make -C /lib/modules/6.13.2-2-cachyos/build M=/maccel/driver
> make[2]: Entering directory '/usr/lib/modules/6.13.2-2-cachyos/build'
> make[3]: Entering directory '/maccel/driver'
> warning: the compiler differs from the one used to build the kernel
>   The kernel was built by: clang version 19.1.7
>   You are using:           gcc (GCC) 14.2.1 20250207
>   CC [M]  maccel.o
> gcc: error: unrecognized command-line option ‘-mretpoline-external-thunk’
> gcc: error: unrecognized command-line option ‘-fsplit-lto-unit’
> gcc: error: unrecognized command-line option ‘-fdebug-info-for-profiling’
> gcc: error: unrecognized command-line option ‘-mllvm’
> gcc: error: unrecognized command-line option ‘-mllvm’
> gcc: error: unrecognized command-line option ‘-improved-fs-discriminator=true’
> gcc: error: unrecognized command-line option ‘-fbasic-block-sections=labels’
> make[5]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/scripts/Makefile.build:196: maccel.o] Error 1
> make[4]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/Makefile:1997: .] Error 2
> make[3]: *** [/usr/lib/modules/6.13.2-2-cachyos/build/Makefile:251: __sub-make] Error 2
> make[3]: Leaving directory /maccel/driver'
> make[2]: *** [Makefile:251: __sub-make] Error 2
> make[2]: Leaving directory '/usr/lib/modules/6.13.2-2-cachyos/build'
> make[1]: *** [Makefile:7: build] Error 2
> make[1]: Leaving directory '/maccel/driver'
> make: *** [Makefile:16: build] Error 2
> ```
> 
Weird. `export CC=clang;` and then run the make commands again.

> It's a Lamzu 4k mini. (select the name, right click, search google for...)
> 
lul, a little sassy, aren't we.

After a restart, I tried doing a `make install` on the branch. But got this error.

```
...
changed ownership of '/sys/module/maccel/parameters/OFFSET' from root:root to :maccel
changed ownership of '/sys/module/maccel/parameters/OUTPUT_CAP' from root:root to :maccel
changed ownership of '/sys/module/maccel/parameters/SENS_MULT' from root:root to :maccel
chown: cannot access '/var/opt/maccel/resets/*': No such file or directory
failed to change ownership of '/var/opt/maccel/resets/*' to :maccel
changed ownership of '/dev/maccel' from root:root to :maccel
make: *** [Makefile:19: install] Error 1
```

Maybe it's because that one commit behind?

```
❯ dkms status | grep maccel                                                                                                                                                                               
maccel/0.4.0, 6.13.4-2-cachyos, x86_64: installed

~/maccel on  feat/support-high-polling-rate
❯ make install_debug                                                                                                                                                                                        ...
changed ownership of '/sys/module/maccel/parameters/SENS_MULT' from root:root to :maccel
chown: cannot access '/var/opt/maccel/resets/*': No such file or directory
failed to change ownership of '/var/opt/maccel/resets/*' to :maccel
changed ownership of '/dev/maccel' from root:root to :maccel
make: *** [Makefile:19: install] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -2,10 +2,15 @@ DRIVERDIR?=`pwd`/driver
 MODULEDIR?=/lib/modules/`uname -r`/kernel/drivers/usb
 
 build:
-	$(MAKE) EXTRA_CFLAGS='$(EXTRA_CFLAGS)' -C $(DRIVERDIR)
+	$(MAKE) EXTRA_CFLAGS="$(EXTRA_CFLAGS)" -C $(DRIVERDIR)
 
 build_debug: EXTRA_CFLAGS = -g -DDEBUG
 build_debug: build
+
+test: 
+	$(MAKE) -C $(DRIVERDIR) test
+test_debug:
+	$(MAKE) -C $(DRIVERDIR) test_debug
 
 install_debug: build_debug install
 

--- a/PKGBUILD
+++ b/PKGBUILD
@@ -1,6 +1,6 @@
 pkgname=maccel-dkms
 _pkgname="maccel"
-pkgver=0.4.0
+pkgver=0.4.1
 pkgrel=1
 pkgdesc="Mouse acceleration driver and kernel module for Linux."
 arch=("x86_64")

--- a/driver/Makefile
+++ b/driver/Makefile
@@ -12,7 +12,7 @@ else
 endif
 
 build:
-	$(MAKE) CC=$(CC) EXTRA_CFLAGS='$(EXTRA_CFLAGS)' -C $(KDIR) M=$(CURDIR)
+	$(MAKE) CC=$(CC) EXTRA_CFLAGS="$(EXTRA_CFLAGS)" -C $(KDIR) M=$(CURDIR)
 
 build_debug: EXTRA_CFLAGS += -g -DDEBUG
 build_debug: build
@@ -20,10 +20,9 @@ build_debug: build
 clean:
 		$(MAKE) -C $(KDIR) M=$(CURDIR) clean
 
-test_debug: **/*.test.c
-	@mkdir -p tests/snapshots
-	@sh tests/run_tests.sh -DDEBUG
+test_debug: EXTRA_CFLAGS += -g -DDEBUG
+test_debug: test
 
 test: **/*.test.c
 	@mkdir -p tests/snapshots
-	@sh tests/run_tests.sh $(DEBUG)
+	EXTRA_CFLAGS="$(EXTRA_CFLAGS)" TEST_NAME=$(name) sh tests/run_tests.sh

--- a/driver/accel.h
+++ b/driver/accel.h
@@ -46,8 +46,7 @@ static inline fixedpt sensitivity(fixedpt input_speed, fixedpt param_sens_mult,
   return sens;
 }
 
-static inline fixedpt input_speed(fixedpt dx, fixedpt dy,
-                                  u32 polling_interval) {
+static inline fixedpt input_speed(fixedpt dx, fixedpt dy, fixedpt time_ms) {
   fixedpt a2 = fixedpt_mul(dx, dx);
   fixedpt b2 = fixedpt_mul(dy, dy);
   fixedpt a2_plus_b2 = fixedpt_add(a2, b2);
@@ -59,22 +58,21 @@ static inline fixedpt input_speed(fixedpt dx, fixedpt dy,
   fixedpt distance = fixedpt_sqrt(a2_plus_b2);
 
   if (distance == -1) {
-    dbg("distance calculation failed: t = %d", polling_interval);
+    dbg("distance calculation failed: t = %s", fptoa(time_ms));
     return 0;
   }
 
   dbg("distance (in)              %s", fptoa(distance));
 
-  fixedpt time = fixedpt_fromint(polling_interval);
-  fixedpt speed = fixedpt_div(distance, time);
+  fixedpt speed = fixedpt_div(distance, time_ms);
 
-  dbg("polling interval           %s", fptoa(time));
+  dbg("time interval              %s", fptoa(time_ms));
   dbg("speed (in)                 %s", fptoa(speed));
 
   return speed;
 }
 
-static inline AccelResult f_accelerate(int x, int y, u32 polling_interval,
+static inline AccelResult f_accelerate(int x, int y, fixedpt time_interval_ms,
                                        fixedpt param_sens_mult,
                                        fixedpt param_accel,
                                        fixedpt param_offset,
@@ -91,7 +89,7 @@ static inline AccelResult f_accelerate(int x, int y, u32 polling_interval,
   dbg("in: x (fixedpt conversion) %s", fptoa(dx));
   dbg("in: y (fixedpt conversion) %s", fptoa(dy));
 
-  fixedpt speed_in = input_speed(dx, dy, polling_interval);
+  fixedpt speed_in = input_speed(dx, dy, time_interval_ms);
   fixedpt sens = sensitivity(speed_in, param_sens_mult, param_accel,
                              param_offset, param_output_cap);
 

--- a/driver/accel.h
+++ b/driver/accel.h
@@ -11,8 +11,6 @@ typedef struct {
   int y;
 } AccelResult;
 
-const fixedpt FIXEDPT_ZERO = fixedpt_rconst(0.0);
-
 /**
  * Calculate the factor by which to multiply the input vector
  * in order to get the desired output speed.
@@ -28,7 +26,7 @@ static inline fixedpt sensitivity(fixedpt input_speed, fixedpt param_sens_mult,
 
   dbg("accel    %s", fptoa(param_accel));
 
-  if (input_speed > FIXEDPT_ZERO) {
+  if (input_speed > 0) {
     sens = fixedpt_add(FIXEDPT_ONE, fixedpt_mul((param_accel), input_speed));
   }
 
@@ -39,7 +37,7 @@ static inline fixedpt sensitivity(fixedpt input_speed, fixedpt param_sens_mult,
   dbg("sens     %s", fptoa(sens));
   dbg("sens cap %s", fptoa(output_cap));
 
-  if (param_output_cap != FIXEDPT_ZERO && sens > output_cap) {
+  if (param_output_cap != 0 && sens > output_cap) {
     return output_cap;
   }
 
@@ -79,8 +77,8 @@ static inline AccelResult f_accelerate(int x, int y, fixedpt time_interval_ms,
                                        fixedpt param_output_cap) {
   AccelResult result = {.x = 0, .y = 0};
 
-  static fixedpt carry_x = FIXEDPT_ZERO;
-  static fixedpt carry_y = FIXEDPT_ZERO;
+  static fixedpt carry_x = 0;
+  static fixedpt carry_y = 0;
 
   fixedpt dx = fixedpt_fromint(x);
   fixedpt dy = fixedpt_fromint(y);

--- a/driver/accelk.h
+++ b/driver/accelk.h
@@ -7,43 +7,61 @@
 #include "params.h"
 
 const fixedpt HUNDRED_MS = fixedpt_rconst(100);
-const fixedpt US_PER_MS = fixedpt_rconst(1000);
 
-const u64 EIGTH_OF_A_MS_IN_US = 125; // To support up to 8000Hz polling rate
-                                     // we should cap the floor to 0.125ms.
-                                     // This is an arbitrary decision...
+// To support up to 8000Hz polling rate
+// we should cap the floor to 0.125ms.
+// This is an arbitrary decision...
+#if FIXEDPT_BITS == 64
+const s64 EIGTH_OF_A_MS_IN_UNIT = 125000; // 0.125ms in ns
+const fixedpt UNIT_PER_MS = fixedpt_rconst(1000000);
+#else
+const s64 EIGTH_OF_A_MS_IN_UNIT = 125; // 0.125ms in us
+const fixedpt UNIT_PER_MS = fixedpt_rconst(1000);
+#endif
 
 static inline AccelResult accelerate(int x, int y) {
   dbg("FIXEDPT_BITS = %d", FIXEDPT_BITS);
 
   static ktime_t last_time;
-  static s64 last_us = EIGTH_OF_A_MS_IN_US;
+  static s64 last_unit_time = EIGTH_OF_A_MS_IN_UNIT;
   ktime_t now = ktime_get();
 
-  s64 us = ktime_to_us(now - last_time);
-
+#if FIXEDPT_BITS == 64
+  s64 unit_time = ktime_to_ns(now - last_time);
+  dbg("ktime interval -> now (%llu) vs last_ktime (%llu), diff = %llins", now,
+      last_time, unit_time);
+#else
+  s64 unit_time = ktime_to_us(now - last_time);
   dbg("ktime interval -> now (%llu) vs last_ktime (%llu), diff = %llius", now,
-      last_time, us);
+      last_time, unit_time);
+#endif
 
   last_time = now;
 
-  if (us < EIGTH_OF_A_MS_IN_US) { // setting a floor for safety
-    us = last_us;
+  if (unit_time < EIGTH_OF_A_MS_IN_UNIT) { // setting a floor for safety
+    unit_time = last_unit_time;
   }
 
-  last_us = us;
+  last_unit_time = unit_time;
 
-  fixedpt _us = fixedpt_fromint(us);
-  fixedpt ms = fixedpt_div(_us, US_PER_MS);
+  fixedpt _unit_time = fixedpt_fromint(unit_time);
+  fixedpt millisecond = fixedpt_div(_unit_time, UNIT_PER_MS);
 
-  dbg("ktime interval, converting to ms: %lluus -> %sms", us, fptoa(ms));
+#if FIXEDPT_BITS == 64
+  dbg("ktime interval, converting to ns: %lluns -> %sms", unit_time,
+      fptoa(millisecond));
+#else
+  dbg("ktime interval, converting to us: %lluus -> %sms", unit_time,
+      fptoa(millisecond));
+#endif
 
-  if (ms > HUNDRED_MS) { // rounding down to 100 ms
-    ms = HUNDRED_MS;
+  if (millisecond > HUNDRED_MS) { // rounding down to 100 ms
+    millisecond = HUNDRED_MS;
   }
 
-  return f_accelerate(x, y, ms, atofp(PARAM_SENS_MULT), atofp(PARAM_ACCEL),
-                      atofp(PARAM_OFFSET), atofp(PARAM_OUTPUT_CAP));
+  return f_accelerate(x, y, millisecond, atofp(PARAM_SENS_MULT),
+                      atofp(PARAM_ACCEL), atofp(PARAM_OFFSET),
+                      atofp(PARAM_OUTPUT_CAP));
 }
 
 #endif // !_ACCELK_H_

--- a/driver/accelk.h
+++ b/driver/accelk.h
@@ -6,24 +6,16 @@
 #include "linux/ktime.h"
 #include "params.h"
 
-const fixedpt HUNDRED_MS = fixedpt_rconst(100);
-
-// To support up to 8000Hz polling rate
-// we should cap the floor to 0.125ms.
-// This is an arbitrary decision...
 #if FIXEDPT_BITS == 64
-const s64 EIGTH_OF_A_MS_IN_UNIT = 125000; // 0.125ms in ns
-const fixedpt UNIT_PER_MS = fixedpt_rconst(1000000);
+const fixedpt UNIT_PER_MS = fixedpt_rconst(1000000); // 1 million nanoseconds
 #else
-const s64 EIGTH_OF_A_MS_IN_UNIT = 125; // 0.125ms in us
-const fixedpt UNIT_PER_MS = fixedpt_rconst(1000);
+const fixedpt UNIT_PER_MS = fixedpt_rconst(1000); // 1 thousand microsends
 #endif
 
 static inline AccelResult accelerate(int x, int y) {
   dbg("FIXEDPT_BITS = %d", FIXEDPT_BITS);
 
   static ktime_t last_time;
-  static s64 last_unit_time = EIGTH_OF_A_MS_IN_UNIT;
   ktime_t now = ktime_get();
 
 #if FIXEDPT_BITS == 64
@@ -35,29 +27,18 @@ static inline AccelResult accelerate(int x, int y) {
   dbg("ktime interval -> now (%llu) vs last_ktime (%llu), diff = %llius", now,
       last_time, unit_time);
 #endif
-
   last_time = now;
-
-  if (unit_time < EIGTH_OF_A_MS_IN_UNIT) { // setting a floor for safety
-    unit_time = last_unit_time;
-  }
-
-  last_unit_time = unit_time;
 
   fixedpt _unit_time = fixedpt_fromint(unit_time);
   fixedpt millisecond = fixedpt_div(_unit_time, UNIT_PER_MS);
 
 #if FIXEDPT_BITS == 64
-  dbg("ktime interval, converting to ns: %lluns -> %sms", unit_time,
+  dbg("ktime interval -> converting to ns: %lluns -> %sms", unit_time,
       fptoa(millisecond));
 #else
   dbg("ktime interval, converting to us: %lluus -> %sms", unit_time,
       fptoa(millisecond));
 #endif
-
-  if (millisecond > HUNDRED_MS) { // rounding down to 100 ms
-    millisecond = HUNDRED_MS;
-  }
 
   return f_accelerate(x, y, millisecond, atofp(PARAM_SENS_MULT),
                       atofp(PARAM_ACCEL), atofp(PARAM_OFFSET),

--- a/driver/accelk.h
+++ b/driver/accelk.h
@@ -2,24 +2,32 @@
 #define _ACCELK_H_
 
 #include "accel.h"
+#include "fixedptc.h"
 #include "linux/ktime.h"
 #include "params.h"
 
 const fixedpt HUNDRED_MS = fixedpt_rconst(100);
 const fixedpt US_PER_MS = fixedpt_rconst(1000);
 
+const u64 EIGTH_OF_A_MS_IN_US = 125; // To support up to 8000Hz polling rate
+                                     // we should cap the floor to 0.125ms.
+                                     // This is an arbitrary decision...
+
 static inline AccelResult accelerate(int x, int y) {
   dbg("FIXEDPT_BITS = %d", FIXEDPT_BITS);
 
   static ktime_t last_time;
-  static s64 last_us = 100; // .1ms
-
+  static s64 last_us = EIGTH_OF_A_MS_IN_US;
   ktime_t now = ktime_get();
+
   s64 us = ktime_to_us(now - last_time);
+
+  dbg("ktime interval -> now (%llu) vs last_ktime (%llu), diff = %llius", now,
+      last_time, us);
 
   last_time = now;
 
-  if (us < 100) { // ensure no less than .1ms?
+  if (us < EIGTH_OF_A_MS_IN_US) { // setting a floor for safety
     us = last_us;
   }
 
@@ -27,6 +35,8 @@ static inline AccelResult accelerate(int x, int y) {
 
   fixedpt _us = fixedpt_fromint(us);
   fixedpt ms = fixedpt_div(_us, US_PER_MS);
+
+  dbg("ktime interval, converting to ms: %lluus -> %sms", us, fptoa(ms));
 
   if (ms > HUNDRED_MS) { // rounding down to 100 ms
     ms = HUNDRED_MS;

--- a/driver/accelk.h
+++ b/driver/accelk.h
@@ -4,27 +4,32 @@
 #include "accel.h"
 #include "linux/ktime.h"
 #include "params.h"
-#include "utils.h"
+
+const fixedpt HUNDRED_MS = fixedpt_rconst(100);
+const fixedpt US_PER_MS = fixedpt_rconst(1000);
 
 static inline AccelResult accelerate(int x, int y) {
   dbg("FIXEDPT_BITS = %d", FIXEDPT_BITS);
 
-  static ktime_t last;
-  static u64 last_ms = 1;
+  static ktime_t last_time;
+  static s64 last_us = 100; // .1ms
 
   ktime_t now = ktime_get();
-  u64 ms = ktime_to_ms(now - last);
+  s64 us = ktime_to_us(now - last_time);
 
-  last = now;
+  last_time = now;
 
-  if (ms < 1) { // ensure no less than 1ms
-    ms = last_ms;
+  if (us < 100) { // ensure no less than .1ms?
+    us = last_us;
   }
 
-  last_ms = ms;
+  last_us = us;
 
-  if (ms > 100) { // rounding dow to 100 ms
-    ms = 100;
+  fixedpt _us = fixedpt_fromint(us);
+  fixedpt ms = fixedpt_div(_us, US_PER_MS);
+
+  if (ms > HUNDRED_MS) { // rounding down to 100 ms
+    ms = HUNDRED_MS;
   }
 
   return f_accelerate(x, y, ms, atofp(PARAM_SENS_MULT), atofp(PARAM_ACCEL),

--- a/driver/input_echo.h
+++ b/driver/input_echo.h
@@ -46,7 +46,8 @@ static ssize_t read(struct file *f, char __user *user_buffer, size_t size,
                     loff_t *offset) {
   int x = MOUSE_MOVE_CACHE[0];
   int y = MOUSE_MOVE_CACHE[1];
-  fixedpt speed = input_speed(fixedpt_fromint(x), fixedpt_fromint(y), 1);
+  fixedpt speed =
+      input_speed(fixedpt_fromint(x), fixedpt_fromint(y), FIXEDPT_ONE);
 
   dbg("echoing speed to userspace: %s", fptoa(speed));
 

--- a/driver/tests/accel.test.c
+++ b/driver/tests/accel.test.c
@@ -15,8 +15,8 @@ static int test_acceleration(const char *filename, fixedpt param_sens_mult,
   for (int x = MIN; x < MAX; x++) {
     for (int y = MIN; y < MAX; y++) {
 
-      result = f_accelerate(x, y, 1, param_sens_mult, param_accel, param_offset,
-                            param_output_cap);
+      result = f_accelerate(x, y, FIXEDPT_ONE, param_sens_mult, param_accel,
+                            param_offset, param_output_cap);
       char curr_debug_print[LINE_LEN];
 
       sprintf(curr_debug_print, "(%d, %d) => (%d, %d)\n", x, y, result.x,

--- a/driver/tests/input_speed.test.c
+++ b/driver/tests/input_speed.test.c
@@ -2,9 +2,10 @@
 #include "./test_utils.h"
 #include <stdio.h>
 
-int assert_string_value(char *filename, double x, double y, double dt) {
+int assert_string_value(char *filename, double x, double y, double t) {
   fixedpt dx = fixedpt_rconst(x);
   fixedpt dy = fixedpt_rconst(y);
+  fixedpt dt = fixedpt_rconst(t);
 
   dbg("in                        (%f, %f)", x, y);
   dbg("in: x (fixedpt conversion) %s", fptoa(x));
@@ -13,10 +14,10 @@ int assert_string_value(char *filename, double x, double y, double dt) {
   fixedpt s = input_speed(dx, dy, dt);
 
   double res = fixedpt_todouble(s);
-  dbg("(%f, %f) dt = %f -> %f\n", x, y, dt, res);
+  dbg("(%f, %f) dt = %f -> %f\n", x, y, t, res);
 
   char content[100];
-  sprintf(content, "(sqrt(%f, %f) / %f) = %f\n", x, y, dt, res);
+  sprintf(content, "(sqrt(%f, %f) / %f) = %f\n", x, y, t, res);
 
   assert_snapshot(filename, content);
 

--- a/driver/tests/run_tests.sh
+++ b/driver/tests/run_tests.sh
@@ -1,5 +1,10 @@
+#!/bin/bash
+
 for test in tests/*.test.c; do
-  gcc ${test} -o maccel_test -g -lm $1
+  if [[ ! $test =~ $TEST_NAME ]]; then
+    continue
+  fi
+  gcc ${test} -o maccel_test -lm $EXTRA_CFLAGS
   ./maccel_test
   rm maccel_test
 done

--- a/driver/tests/test_utils.h
+++ b/driver/tests/test_utils.h
@@ -8,8 +8,6 @@
 #include <sys/wait.h>
 #include <unistd.h>
 
-typedef int bool;
-
 static int diff(const char *content, const char *filename) {
 
   int pipe_fd[2];
@@ -84,7 +82,7 @@ static void assert_snapshot(const char *__filename, const char *content) {
     exit(1);
   }
 
-  bool snapshot_file_exists = access(filename, F_OK) != -1;
+  int snapshot_file_exists = access(filename, F_OK) != -1;
   FILE *snapshot_file;
 
   if (snapshot_file_exists) {
@@ -105,7 +103,7 @@ static void assert_snapshot(const char *__filename, const char *content) {
 
     stat(filename, &stats);
     file_size = stats.st_size;
-    char *snapshot = malloc(stats.st_size + 1);
+    char *snapshot = (char *)malloc(stats.st_size + 1);
 
     if (snapshot == NULL) {
       fprintf(stderr,
@@ -120,14 +118,15 @@ static void assert_snapshot(const char *__filename, const char *content) {
       fprintf(stderr, "failed to read a snapshot file %s\n", filename);
       exit(1);
     }
+    snapshot[file_size] = 0; // null byte terminator
 
-    int string_test = strcmp(snapshot, content);
+    int string_test_diff = strcmp(snapshot, content);
 
     diff(content, filename);
 
     /* dbg("diff in content = %d: snapshot '%s' vs now '%s'", string_test, */
     /*     snapshot, content); */
-    assert(string_test == 0);
+    assert(string_test_diff == 0);
   } else {
     fprintf(snapshot_file, "%s", content);
     printf("created a snapshot file %s\n", filename);

